### PR TITLE
feat: log gradient norm during training

### PR DESCRIPTION
**Pull request by gabrielfruet**
## What has changed and why?

Logs the total gradient norm (`gradient_norm`) during finetuning (task path). `TrainModel.clip_gradients` now returns the pre-clipping total norm; the train loop logs it as `gradient_norm` (shown in the console as `grad_norm` and written to all configured loggers: JSONL, TensorBoard, Weights & Biases, MLflow). When gradient clipping is disabled, the norm is computed via an unbounded (`inf`) norm — a single pass over the parameters.

Closes TRN-2254.

Scope: finetuning/task path only. The pretrain path (`_methods/*` `configure_gradient_clipping`) is out of scope for a follow-up.

## How has it been tested?

- New `tests/_task_models/test_train_model.py` — verifies the returned norm matches an independent L2 computation, for both enabled (`gradient_clip_val > 0`) and disabled clipping; no mocking.
- New `log_step` tests in `tests/_commands/test_train_task_helpers.py` — `grad_norm` shown when present, omitted when `None`.
- `ruff format`/`check` and `mypy` clean on changed files.
- Regression sweep (`tests/_task_models` + `test_train_task_helpers` + `test_train_helpers`): 323 passed, 4 skipped, 1 xfailed.
- End-to-end: confirmed `gradient_norm` written to the TensorBoard scalar and shown in `train.log` over a short CPU run.

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [x] Yes

## Did you update the documentation?

- [x] Yes — added a "Logged metrics" section to `docs/source/settings/train_settings.md`.

### CODE REVIEW

File: src/lightly_train/_commands/train_task.py [hunk 1]
```diff
@@ -1669,6 +1675,11 @@ def _train_task_from_config(config: TrainTaskConfig) -> None:
                     global_batch_size=effective_global_batch_size,
                     gradient_accumulation_steps=config.gradient_accumulation_steps,
                     learning_rate=current_lr,
+                    gradient_norm=(
+                        float(gradient_norm.detach())
+                        if gradient_norm is not None
+                        else None
+                    ),
                 )
```
Comment (mrpositron) [inline src/lightly_train/_commands/train_task.py:1682]: `float(gradient_norm.detach())` conversion is done twice here, check line 1665. you can do smth like this:
```python
grad_norm_value = float(gradient_norm.detach()) if gradient_norm is not None else None
if grad_norm_value is not None:
    train_log_dict["gradient_norm"] = grad_norm_value
...
log_step(..., gradient_norm=grad_norm_value)
```

File: tests/_task_models/test_train_model.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    returned_norm = train_model.clip_gradients(fabric=fabric, optimizer=optimizer)
+
+    assert returned_norm is not None
+    assert torch.allclose(returned_norm, expected_norm, rtol=1e-5), (
+        f"Expected {expected_norm}, got {returned_norm}"
+    )
+
+
+def test_clip_gradients__returns_none_when_no_gradients() -> None:
```
Comment (mrpositron) [inline tests/_task_models/test_train_model.py]: Confusing test function naming, `test_clip_gradients__returns_none_when_no_gradients`. Asserts the method returns a (zero) tensor, not `None`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,11 @@ this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ### Added
 
+- Log the total gradient norm (`gradient_norm`) during finetuning. It is shown in the
+  console as `grad_norm` and written to all configured loggers (JSONL, TensorBoard,
+  Weights & Biases, MLflow). It is the pre-clipping norm when gradient clipping is
+  enabled and an unbounded norm otherwise.
```
Comment (mrpositron) [inline CHANGELOG.md]: "unbounded norm" might be confusing with infinity norm ($\mathcal{L}_{\infty}$).

File: docs/source/settings/train_settings.md [hunk 4]
```diff
@@ -594,6 +594,21 @@ lightly_train.train_object_detection(
...
 
+#### Logged metrics
+
+In addition to task-specific metrics, LightlyTrain logs the following values for
+training steps, at the cadence set by [`log_every_num_steps`](#log_every_num_steps):
+
+- `gradient_norm`: Total gradient norm computed after backpropagation, before the
+  optimizer step. If gradient clipping is enabled (`gradient_clip_val > 0`) this is the
+  pre-clipping norm; otherwise it is computed via an unbounded (`inf`) norm. Use it to
+  spot exploding or vanishing gradients during training. It is also shown in the console
```
Comment (mrpositron) [inline docs/source/settings/train_settings.md]: "unbounded norm" might be confusing with infinity norm ($\mathcal{L}_{\infty}$).

File: src/lightly_train/_task_models/image_classification/train_model.py [hunk 5]
```diff
@@ -296,14 +296,15 @@ def set_train_mode(self) -> None:
...
-        if no_auto(self.model_args.gradient_clip_val) > 0:
-            fabric.clip_gradients(
-                module=self,
-                optimizer=optimizer,
-                max_norm=no_auto(self.model_args.gradient_clip_val),
-                error_if_nonfinite=False,
-            )
+    def clip_gradients(self, fabric: Fabric, optimizer: Optimizer) -> Tensor | None:
+        gradient_clip_val = no_auto(self.model_args.gradient_clip_val)
+        max_norm = gradient_clip_val if gradient_clip_val > 0 else float("inf")
```
[the remark this answers is not in the document]
Comment (gabrielfruet) [reply src/lightly_train/_task_models/image_classification/train_model.py]: Fixed in feccfc0b. You were right — with a non-finite grad, `clip_grad_norm_(max_norm=inf)` produces `clip_coef = inf / inf = nan` and multiplies every gradient by nan. Confirmed empirically on torch 2.8.0.

Fix: when `gradient_clip_val <= 0`, `clip_gradients` now computes the total norm via a new `TrainModel._total_gradient_norm()` helper that does NOT touch gradients (`sum(p.grad.pow(2).sum()).sqrt()` over params with grad). The enabled branch is unchanged and still returns the pre-clipping norm from `fabric.clip_gradients`.

Applied to all 12 overriding subclasses (same latent pattern), not just image_classification. Added `test_clip_gradients__disabled_does_not_mutate_gradients` which injects an inf grad, calls `clip_gradients`, and asserts (a) every gradient is unchanged and (b) the returned norm is non-finite.

File: src/lightly_train/_commands/train_task.py (hunk 1 above)
Comment (gabrielfruet) [reply src/lightly_train/_commands/train_task.py:1682]: Fixed in feccfc0b. Hoisted the conversion into a single `grad_norm_value` at the top of the log block:

```
grad_norm_value = float(gradient_norm.detach()) if gradient_norm is not None else None
if grad_norm_value is not None:
    train_log_dict["gradient_norm"] = grad_norm_value
...
helpers.log_step(..., gradient_norm=grad_norm_value)
```

So `float(gradient_norm.detach())` now appears exactly once in the file.

File: tests/_task_models/test_train_model.py (hunk 2 above)
Comment (gabrielfruet) [reply tests/_task_models/test_train_model.py]: Fixed in feccfc0b. Renamed to `test_clip_gradients__returns_zero_norm_when_no_gradients` and corrected the docstring (the old one incorrectly referenced a nonexistent `Fabric.clip_gradients_norm`). The assertion still expects a zero tensor, matching the new name.

File: CHANGELOG.md (hunk 3 above)
Comment (gabrielfruet) [reply CHANGELOG.md]: Fixed in feccfc0b. Reworded "an unbounded norm otherwise" → "the total gradient norm computed without clipping otherwise". This also matches the new implementation: we now compute the norm non-mutatingly instead of via `clip_grad_norm_(max_norm=inf)`.

File: docs/source/settings/train_settings.md (hunk 4 above)
Comment (gabrielfruet) [reply docs/source/settings/train_settings.md]: Fixed in feccfc0b. Reworded "computed via an unbounded (`inf`) norm" → "the total gradient norm computed without applying gradient clipping". This avoids the L∞ confusion and matches the new non-mutating implementation. Same wording applied to CHANGELOG.md.

File: src/lightly_train/_task_models/train_model.py [hunk 6]
```diff
@@ -114,8 +115,25 @@ def get_export_state_dict(self) -> dict[str, Any]:
...
+        # Clip the gradients. Returns the total gradient norm before clipping if
+        # the norm-based clipping algorithm is used, otherwise None. Returning None
+        # here means that the gradient norm is not logged for this model.
+        return None
+
+    def _total_gradient_norm(self) -> Tensor:
+        # Total L2 gradient norm across all parameters with a grad, computed
+        # without clipping or otherwise mutating the gradients. Returns 0.0
+        # when no parameter currently has a gradient.
+        total_sq = torch.zeros(1)
```
Comment (mrpositron) [inline src/lightly_train/_task_models/train_model.py]: `_total_gradient_norm()` initializes `total_sq = torch.zeros(1)` on CPU, then adds `p.grad` sums to it. On GPU training with clipping disabled, this will mix CPU and accelerator tensors and raise a device mismatch before the optimizer step.

Comment (gabrielfruet) [reply src/lightly_train/_task_models/train_model.py]: Addressed on fa00bd6

**Review by mrpositron — APPROVED** [review]
LGTM 🚀, after CI issues resolved
